### ISSUE TITLE
Distributed atomic counter support in JetStream KV

### ISSUE THREAD

**Issue by FalxDrako** [nats-io/nats-server#2656]
## Feature Request

Add Distributed Atomic Counter support to **JetStream KV**.

#### Use Case:

 [Slack thread](https://natsio.slack.com/archives/CM3T6T7JQ/p1635605994108300)

#### Proposed Change:

- add supporting code to **kv**
- add command-line support
   - a new flag for `kv add` subcommand to specify whether the stream backing a bucket should be a **file stream** or a **memory stream**, e.g.:  
     - `nats kv add my_bucket --replicas 1 --history 5 --STORAGE=FILE` 
     - `nats kv add my_bucket --replicas 1 --history 5 --STORAGE=MEMORY`
    - two new subcommands:
      - `incrby x`, e.g.: `nats kv INCRBY 20 my_bucket my_key_1`
      - `decrby x`, e.g.: `nats kv DECRBY 10 my_bucket my_key_1`
- allow interacting with the counters **programmatically**
  - **1st phase**: via publishing messages to dedicated subjects; this would allow the new feature to be used before the official clients catch up.
  - **2nd phase**: via Synadia official JetStream client(s).

#### Who Benefits From The Change(s)?

- **Synadia** - by closing the feature gap between JetStream and more popular solutions.
- **Synadia** - by making clients and users happy. Who doesn't need or use counters in production?
- **JetSstream** users - who can say no to happiness ?

#### Alternative Approaches

- The competition
  - Apache Kafka
  - Apache Pulsar
- Other software
  - Redis
  - RDBMS
  - etc

**Comment by ajlane** [comment]
The Slack thread describing the use case has been archived, but I hope that the use cases are still easy enough to imagine.

I am currently doing this this by publishing +1s and -1s to a stream, then consuming that stream with another client that _eventually_ puts the accumulated results into a kv store. It would be nice to skip that step.

**Comment by elichai** [comment]
Is anyone still working/thinking about this?

**Comment by ripienaar** [comment]
it's on the short term roadmap yes

**Comment by pySilver** [comment]
@ripienaar chances this will be in 2.11?

**Comment by ripienaar** [comment]
I believe it's still the plan yes

**Comment by sloveridge** [comment]
I have been checking for this in the 2.11 preview announcements but haven't seen it land yet. Could you let us know if this will be in 2.11 still?

Thanks you!

**Comment by derekcollison** [comment]
Yes it will be. We are admittedly behind, but not for lack of effort and commitment, we just been helping out customers and partners. But it will be there.

**Comment by Ilyes512** [comment]
Any updates on this? Is this tight to a specific issue in the 2.11 milestone that we could track?

**Comment by derekcollison** [comment]
No updates but still planned, just running a bit behind.

**Comment by max13fr** [comment]
Hello,

Does anyone have found a workaround to wait this feature ?

We can read the value then update with the incremented value, but I didn't find any transactional command that avoid missing an increment if 2 get values are done before any update (so only incremented by one instead of two).

Thanks in advance,
Best regards, Max

**Comment by pySilver** [comment]
It will not be incremented by two because there is versioning

**Comment by Jarema** [comment]
@max13fr you can use `kv.Update()` method, which takes a revision as an argument (that you get with the kv value). It would fail if you missed an increment.

**Comment by MauriceVanVeen** [comment]
This has been implemented and will be released as part of 2.12 due in a few months 🎉 
We'll likely do a preview release earlier than that, so stay tuned!

Design doc: [JetStream Distributed Counter CRDT](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-49.md)

Relevant server PRs: https://github.com/nats-io/nats-server/pull/6973, https://github.com/nats-io/nats-server/pull/6988

**Comment by MauriceVanVeen** [comment]
Closing this issue, since it's implemented. Feel free to try it out and open another issue with any feedback/bugs/etc.
